hw-4/data-access/group.repository.ts

```diff
@@ -0,0 +1,79 @@
+import { Group, UserGroup } from '@app-models/index';
+import uuid from 'uuid';
+import sequelize from '../sequelize/sequalize';
+
+
+export default class GroupRepository {
+  private readonly groupModel = Group;
+
+  private readonly userGroupModel = UserGroup;
+
+  getGroups(): Promise<Group[]> {
+    return this.groupModel.findAll();
+  }
+
+  getGroupById(id: string): Promise<Group> {
+    return this.groupModel.findOne({
+      where: {
+        id,
+      },
+    });
+  }
+
+  getGroupByName(name: string): Promise<Group|null> {
+    return this.groupModel.findOne({
+      where: {
+        name,
+      },
+    });
+  }
+
+  add(group: Group): Promise<Group> {
+    const { name, permissions } = group;
+
+    return this.groupModel.create<Group>({
+      id: uuid(),
+      name,
+      permissions,
+    }, {});
+  }
+
+  update(group: Group): Promise<Group> {
+    const {
+      id, name, permissions,
+    } = group;
+
+    return this.groupModel.update({
+      name,
+      permissions,
+    }, {
+      where: {
+        id,
+      },
+      returning: true,
+    }).then(([number, updatedGroups]) => number && updatedGroups[0]);
+  }
+
+  remove(id: string): Promise<number> {
+    return this.groupModel.destroy({
+      where: {
+        id,
+      },
+    });
+  }
+
+  addUsersToGroup(groupId: string, userIds: string[]): Promise<number> {
```

Comment:
We could be add a bit of robustness to the return type and specify each of the operation status.

We would need a new return type for it, something like:

```typescript
type BulkOperationResult = {
   id: string,
  success: Boolean
};
```

Our return type then would be `Promise<BulkOperationResult[]>` where each item would have a `userId` and its operation result status.

Reply:
But do we need such information? I supposed in case of transactions if at least one error appears (success = false) than entire transaction will be rolled back? If so, then we'll always have *sucess* property either true or false for all properties.

Reply:
You're absolutely right. It's been a while I don't work with relational databases and didn't pay attention to the transaction :/